### ISSUE TITLE
[BUG] Handle `opaque` in parsing qasm2 strings 

### ISSUE THREAD

**Issue by TheGupta2012** [qBraid/pyqasm#54]
### Environment

- **PyQASM version**: 0.0.2
- **Python version**: 3.11.6
- **Operating system**: macOS 14.4 

### What happened?

PyQASM should gracefully parse the `opaque` identifier - 

```python 
ValidationError           Traceback (most recent call last)
Cell In[6], line 1
----> 1 mod = pyqasm.load("""
      2 OPENQASM 2.0;
      3 include "qelib1.inc";
      4 qreg q[2];
      5 creg c[2];
      6 opaque custom_gate (a,b,c) p,q,r;""")

File ~/Desktop/qBraid/repos/pyqasm/pyqasm/entrypoint.py:47, in load(program)
     45         program = openqasm3.parse(program)
     46     except openqasm3.parser.QASM3ParsingError as err:
---> 47         raise ValidationError(f"Failed to parse OpenQASM string: {err}") from err
     48 elif not isinstance(program, openqasm3.ast.Program):
     49     raise TypeError("Input quantum program must be of type 'str' or 'openqasm3.ast.Program'.")

ValidationError: Failed to parse OpenQASM string:
```

### Suggestions (Optional)

Handle `opaque` in the visitor

**Comment by ryanhill1** [comment]
Still reproducible on **pyqasm 1.0.4** (`main` @ 5dcae69), and we hit it in production-shaped work rather than synthetically, so adding evidence and a correction to the diagnosis.

### The suggestion in the description won't fix it

> Handle `opaque` in the visitor

The visitor never runs. `opaque` is **OpenQASM 2.0** syntax that was removed in OpenQASM 3, and pyqasm routes qasm2 through the `openqasm3` parser, so this dies at parse time:

```python
from openqasm3 import parse
parse('OPENQASM 2.0;\nopaque Rz(lam) q;\nqreg q[1];\n')   # QASM3ParsingError
parse('OPENQASM 2.0;\ngate mygate(a) q { rz(a) q; }\n')   # parses fine
```

The string `opaque` appears **nowhere** in the installed `openqasm3` package — there is no grammar production for it. So the fix has to live in pyqasm's own preprocessing (`preprocess.py` already rewrites source text to inline includes) or upstream in the grammar, not in the visitor.

### Real-world driver: Quantinuum's `hqslib1.inc`

This blocks parsing any compiled program from Quantinuum H-series hardware. `hqslib1.inc` (shipped in `pytket/qasm/includes/`) is the vendor's qasm2 include file, and it opens with six `opaque` declarations — they're the machine primitives, so they intentionally have no body:

```qasm
opaque Rz(lam) q;
opaque U1q(theta, phi) q;
opaque ZZ() q1,q2;
opaque RZZ(theta) q1,q2;
opaque Rxxyyzz(alpha, beta, gamma) q1,q2;
opaque Rxxyyzz_zphase(alpha, beta, gamma, z0, z1) q1,q2;
```

A compiled bell state on H-series looks like this, and currently cannot be loaded:

```qasm
OPENQASM 2.0;
include "hqslib1.inc";

qreg q[2];
creg c[2];
rz(1.0*pi) q[0];
U1q(0.5*pi,0.5*pi) q[0];
RZZ(0.5*pi) q[0],q[1];
measure q[0] -> c[0];
```

### Confirmation that `opaque` is the only blocker

Replacing the six `opaque` declarations with real gate bodies makes the whole thing work:

```
LOADED OK — qubits: 2 | clbits: 2 | has_measurements: True | depth: 28
```

So custom-include resolution, the gate table, and validation all handle this file correctly. `opaque` is the single point of failure.

### Design note for whoever picks this up

An opaque gate has no decomposition by definition, so `unroll()` semantics need a decision. For `validate()` / `has_measurements()` / rendering, declared arity is sufficient — treating it as a declared-but-undefined black-box gate (similar to the existing external-gate handling) would unblock those without inventing a decomposition.

Two adjacent problems found while working on this, filed separately: #368 (`loads()` skips include resolution, so the string path fails even once `opaque` is supported) and #369 (indirect gate recursion blows the Python stack — you will hit this immediately if you try to fix `opaque` by giving the primitives bodies, because `hqslib1` defines `U` and `CX` in terms of them).

**Comment by ryanhill1** [comment]
Superseded by #370, which carries the corrected diagnosis (this fails at parse time in the `openqasm3` grammar, so the suggestion above to handle it in the visitor cannot reach it), a current reproduction on 1.0.4, and the Quantinuum `hqslib1.inc` driver. Closing as a duplicate — please follow #370.
